File: src/动态规划/template/背包/树上背包/P2014_CTSC1997_选课/Main.php
<?php
const N = 305;
$n = 0;
$m = 0;
$e = array_fill(0, N, []);
$w = array_fill(0, N, 0);
$f = array_fill(0, N, array_fill(0, N, 0));
$siz = array_fill(0, N, 0);

function dfs($u) {
    global $e, $w, $f, $siz, $m;
    $f[$u][1] = $w[$u];
    $siz[$u] = 1;
    foreach ($e[$u] as $v) {
        dfs($v);
        $siz[$u] += $siz[$v];
        for ($j = min($m + 1, $siz[$u]); $j > 0; $j--) {
            for ($k = 0; $k <= min($j - 1, $siz[$v]); $k++) {
                $f[$u][$j] = max($f[$u][$j], $f[$u][$j - $k] + $f[$v][$k]);
            }
        }
    }
}

$input = explode(" ", trim(fgets(STDIN)));
$n = (int)$input[0];
$m = (int)$input[1];

for ($i = 1; $i <= $n; $i++) {
    $input = explode(" ", trim(fgets(STDIN)));
    $k = (int)$input[0];
    $w[$i] = (int)$input[1];
    $e[$k][] = $i;
}

dfs(0);
echo $f[0][$m + 1];
?>

File: src/动态规划/template/背包/树上背包/P2014_CTSC1997_选课/Main.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(100000)

def dfs(u):
    f[u][1] = w[u]
    for v in e[u]:
        dfs(v)
        for j in range(m+1,0,-1):
            for k in range(j):
                f[u][j] = max(f[u][j], f[u][j - k] + f[v][k])

n, m = map(int, input().split())
w = [0] * 305
e = [[] for _ in range(n + 1)]
f = [[0] * 305 for _ in range(305)]
for i in range(1, n+1):
    k, p = map(int, input().split())
    w[i] = p
    e[k].append(i)
dfs(0)
print(f[0][m+1])
